src/lib.rs

```diff
@@ -1,19 +1,137 @@
-//! [Balena](https://www.balena.io) TEMplating ENgine
+//! [balena] **tem**plating **en**gine
 //!
 //! A crate that provides facilities to:
 //!
-//! * evaluate expressions
-//! * evaluate logical expressions
-//! * retrieve AST of any expression for further validation
+//! * parse an expression
+//! * evaluate an expression
+//! * evaluate an expression as a boolean (logical expressions)
+//! * register custom functions and filters
+//!
+//! # Versioning
+//!
+//! This crate is being actively developed and it does NOT follow [Semantic Versioning] yet.
+//! It will follow semantic versioning when it reaches version 1.0.
+//!
+//! MINOR version changes denotes incompatible API changes and PATCH version changes denotes
+//! both new functionality in a backwards-compatible manner and backwards-compatible bug fixes.
+//!
+//! # Examples
+//!
+//! ## Expression parsing
+//!
+//! ```rust
+//! use balena_temen::ast::*;
+//!
+//! let parsed: Expression = "1 + 2 == 3".parse().unwrap();
+//!
+//! let lhs = Expression::new(
+//!     ExpressionValue::Math(
+//!         MathExpression::new(
+//!             Expression::new(ExpressionValue::Integer(1)),
+//!             Expression::new(ExpressionValue::Integer(2)),
+//!             MathOperator::Addition
+//!         )
+//!     )
+//! );
+//! let rhs = Expression::new(
+//!     ExpressionValue::Integer(3)
+//! );
+//! let manual = Expression::new(
+//!     ExpressionValue::Logical(
+//!         LogicalExpression::new(
+//!             lhs, rhs, LogicalOperator::Equal
+//!         )
+//!     )
+//! );
+//! assert_eq!(parsed, manual);
+//! ```
+//!
+//! Visit [ast] module documentation for more info.
+//!
+//! ## Expression evaluation
+//!
+//! ```rust
+//! use balena_temen::{
+//!   ast::Identifier,
+//!   Engine, Context, Value
+//! };
+//! use serde_json::json;
+//!
+//! let engine = Engine::default();       // Default functions, filters
```

Comment:
would there be a way to hide these details from the user in these simple cases ? E.g. to have an API where you do `engine.eval()` and get the result ? Without passing in the `position` and `data` and `context` ? Like a wrapper for simpler cases ? May help with the documentation and adoption, people playing with it and then gradually discovering the need for the other stuff. Then the docs could have an introductory section and then advanced section, and then this example could have two versions for the reader to compare, one per each section. What do you think ?

Reply:
Yup. What's there currently is low level stuff. Helpers (high level) are in the queue. See #20.